### PR TITLE
Correct and improve dev-guide section on fixing graphics-tests.

### PR DESCRIPTION
**Pull request by pp-mo**
Hoping this is worth it, and that this actually is an improvement ..

I *think* this is needed, as I had to work out the fix process for myself, yet again, in course of #3682.

### CODE REVIEW

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 1]
```diff
@@ -75,43 +75,54 @@ If you notice that a graphics test in the Iris testing suite has failed
...
+   process until all graphical tests pass.
+
+#. To add your changes to Iris, you need to make two pull requests :
+
+   * (A) The first PR is made in the test-images-scitools repository, at
+     https://github.com/SciTools/test-images-scitools.  This should contain all
+     the newly-generated PNG files, which must be added to the ``images/v4``
+     directory.  In your Iris repo, these files are to be found in the
+     temporary results folder ``iris/tests/result_image_comparison``.
+     **Note**: this location is covered by a project ``.gitignore`` setting,
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply docs/iris/src/developers_guide/graphics_tests.rst]: Maybe just "this folder" - would that do ?

[the remark this answers is not in the document]
Comment (trexfeathers) [reply docs/iris/src/developers_guide/graphics_tests.rst]: Providing you maintain the distinct uses of "directory" and "folder" in the paragraph, that should work. I personally would have explicitly said "the `result_image_comparison` folder is covered...", but I'm not precious about it!

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 2]
```diff
@@ -75,43 +75,54 @@ If you notice that a graphics test in the Iris testing suite has failed
...
+   relevant test should pass.  However, certain tests do perform *multiple*
+   graphics checks within a single test function : In those cases, any failing
+   check will prevent the others from being run, so a test re-run can encounter
+   further (new) graphical test failures.  Simply repeat the check-and-accept
+   process until all graphical tests pass.
+
+#. To add your changes to Iris, you need to make two pull requests :
+
+   * (A) The first PR is made in the test-images-scitools repository, at
+     https://github.com/SciTools/test-images-scitools.  This should contain all
```
Comment (trexfeathers) [inline docs/iris/src/developers_guide/graphics_tests.rst]: Just visited this repository and it claims to be deprecated in favour of:
https://github.com/SciTools/test-iris-imagehash

Comment (pp-mo) [reply docs/iris/src/developers_guide/graphics_tests.rst]: :+1: Good spot !
Will fix.

**Comment by pp-mo** [comment]
Hi @trexfeathers Given your above comments, it seems it needs a slightly more holistic content review, i.e. what this is, whether it is up-to-date + says the right things.
I'll take another look.

File: docs/iris/src/developers_guide/graphics_tests.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (pp-mo) [reply docs/iris/src/developers_guide/graphics_tests.rst]: :+1:

**Comment by pp-mo** [comment]
Hi @trexfeathers thanks for your attention.
Following your comments above, I've now reviewed the whole page and made quite a few changes.

Your point about this page not appearing on a TOC I think is not a particular problem..
The 'graphics_test' page is linked from a  "graphics tests" section in the `tests.rst` overview page, as a "for-more-detail" reference.  I think that in itself is ok -- it doesn't matter that it does not appear in an index ?
I check that you *can* still find it in searches.

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 3]
```diff
@@ -10,9 +10,10 @@ For this, a basic 'graphics test' assertion operation is provided in the method
 match against a stored reference.
 A "graphics test" is any test which employs this.
 
-At present (Iris version 1.10), such tests include the testing for modules
-`iris.tests.test_plot` and `iris.tests.test_quickplot`, and also some other
-'legacy' style tests (as described in :ref:`developer_tests`).
+At present, such tests include the testing for modules `iris.tests.test_plot`
+and `iris.tests.test_quickplot`, all output plots from the gallery examples
+(contained in `docs/iris/example_tests`), and a few  other 'legacy' style tests
+(as described in :ref:`developer_tests`).
```
Comment (trexfeathers) [inline docs/iris/src/developers_guide/graphics_tests.rst]: Where in `tests.rst` are the legacy style tests described? I can't find any reference to them there.

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 4]
```diff
@@ -38,80 +39,110 @@ Testing actual plot results introduces some significant difficulties :
...
-From Iris v1.11 onward, we want to support testing Iris against multiple
-versions of matplotlib (and some other dependencies).  
-To make this manageable, we have now rewritten "check_graphic" to allow
-multiple alternative 'correct' results without including many more images in
-the Iris repository.  
 This consists of :
 
- * using a perceptual 'image hash' of the outputs (see
-   https://github.com/JohannesBuchner/imagehash) as the basis for checking
+ * The 'check_graphic' funciton uses a perceptual 'image hash' of the outputs
```
Comment (trexfeathers) [inline docs/iris/src/developers_guide/graphics_tests.rst]:
```suggestion
 * The 'check_graphic' function uses a perceptual 'image hash' of the outputs
```

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 5]
```diff
@@ -38,80 +39,110 @@ Testing actual plot results introduces some significant difficulties :
...
 
-Note: the Iris pull-request will not test out successfully in Travis until the
-test-images-scitools pull request has been merged :  This is because there is
-an Iris test which ensures the existence of the reference images (uris) for all
-the targets in the image results database.
 
+How to Add New Known-Valid Result Images
+========================================
+
+When you find that a graphics test in the Iris testing suite has failed,
```
Comment (trexfeathers) [inline docs/iris/src/developers_guide/graphics_tests.rst]: Does this process also work for a completely new test? Feels like the subtitle says so but paragraph seems to be exclusively for failures.

File: docs/iris/src/developers_guide/graphics_tests.rst (hunk 3 above)
Comment (pp-mo) [reply docs/iris/src/developers_guide/graphics_tests.rst]: Right at the top it says "There are three categories of tests within Iris: / * Unit tests / * Integration tests / * Legacy tests"

File: docs/iris/src/developers_guide/graphics_tests.rst (hunk 5 above)
Comment (pp-mo) [reply docs/iris/src/developers_guide/graphics_tests.rst]: The title is supposed to indicate that is about adding new "acceptable" result graphics for an existing testcase.  Which you will only do when the (existing) test fails for some reason.
Writing a whole *new* graphics testcase is somewhat trickier.  I decided not to explain that, after all, as we really don't want any more graphics tests unless absolutely necessary !  
It should probably only be done by people smart enough to work it out for themselves 😈

Comment (pp-mo) [reply docs/iris/src/developers_guide/graphics_tests.rst]: By all means suggest improvements to the words so it's clearer ...
? Please ?

File: docs/iris/src/developers_guide/graphics_tests.rst (hunk 3 above)
Comment (trexfeathers) [reply docs/iris/src/developers_guide/graphics_tests.rst]: Indeed, however the only subsequent mention in that page is how legacy tests should be avoided. If there are any legacy style tests that use graphics tests they are not described in the page.

File: docs/iris/src/developers_guide/graphics_tests.rst (hunk 5 above)
Comment (trexfeathers) [reply docs/iris/src/developers_guide/graphics_tests.rst]: Like the logic, I'll have a go at some appropriate language. Stand by...

File: docs/iris/src/developers_guide/graphics_tests.rst (hunk 3 above)
Comment (pp-mo) [reply docs/iris/src/developers_guide/graphics_tests.rst]: Actually, what "legacy" means in this case is just "before we defined a separation between integration and unit tests".  That is, all the "iris/tests/test_*.py".
The implication is, we can't be bothered to remove/restructure all that, but we don't want any more like it.

I'm basically just not convinced that trying to spell that out will add anything very worthwhile ...

Comment (trexfeathers) [reply docs/iris/src/developers_guide/graphics_tests.rst]: OK. There doesn't seem to be a way out of this that isn't confusing.

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 6]
```diff
@@ -38,80 +39,110 @@ Testing actual plot results introduces some significant difficulties :
 Graphics Testing Strategy
 =========================
 
-Prior to Iris 1.10, all graphics tests compared against a stored reference
-image with a small tolerance on pixel values.
+In the Iris Travis matrix, and over time, graphics tests must run with
+multiple versions of Python, and of key dependencies such as matplotlib.
+To make this manageable, the "check_graphic" test routine tests against
+multiple alternative 'correct' results.  It does this using an image "hash"
```
Comment (trexfeathers) [inline docs/iris/src/developers_guide/graphics_tests.rst]:
```suggestion
multiple alternative 'acceptable' results.  It does this using an image "hash"
```

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 7]
```diff
@@ -38,80 +39,110 @@ Testing actual plot results introduces some significant difficulties :
...
-#. To add your changes to Iris, you need to make two pull requests.  The first
-   should be made to the test-images-scitools repository, and this should
-   contain all the newly-generated png files copied into the folder named
-   'image_files'.
-#. The second pull request should be created in the Iris repository, and should
-   only include the change to the image results database
-   (``tests/results/imagerepo.json``) :
-   This pull request must contain a reference to the matching one in
-   test-images-scitools.
+ * The hashes of 'known accepted results' for each test are stored in a
```
Comment (trexfeathers) [inline docs/iris/src/developers_guide/graphics_tests.rst]:
```suggestion
 * The hashes of known 'acceptable' results for each test are stored in a
```

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 8]
```diff
@@ -38,80 +39,110 @@ Testing actual plot results introduces some significant difficulties :
...
+   lookup dictionary, saved to the repo file
+   ``lib/iris/tests/results/imagerepo.json`` .
+ * An actual reference image for each hash value is stored in a *separate*
+   public repository : https://github.com/SciTools/test-iris-imagehash .
+ * The reference images allow human-eye assessment of whether a new output is
+   judged to be 'close enough' to the older ones, or not.
+ * The utility script ``iris/tests/idiff.py`` automates checking, enabling the
+   developer to easily compare proposed new 'correct' result images against the
```
Comment (trexfeathers) [inline docs/iris/src/developers_guide/graphics_tests.rst]:
```suggestion
   developer to easily compare proposed new 'acceptable' result images against the
```

File: docs/iris/src/developers_guide/graphics_tests.rst [hunk 9]
```diff
@@ -38,80 +39,110 @@ Testing actual plot results introduces some significant difficulties :
...
+   existing accepted reference images, for each failing test.
 
-Note: the Iris pull-request will not test out successfully in Travis until the
-test-images-scitools pull request has been merged :  This is because there is
-an Iris test which ensures the existence of the reference images (uris) for all
-the targets in the image results database.
 
+How to Add New Known-Valid Result Images
```
Comment (trexfeathers) [inline docs/iris/src/developers_guide/graphics_tests.rst]:
```suggestion
How to Add New 'Acceptable' Result Images to Existing Tests
```

File: docs/iris/src/developers_guide/graphics_tests.rst (hunk 5 above)
Comment (trexfeathers) [reply docs/iris/src/developers_guide/graphics_tests.rst]: See new review

**Comment by trexfeathers** [comment]
@pp-mo I believe if you rebase this onto master the tests will also pass again

**Comment by pp-mo** [comment]
Rebased ... :sunglasses:

**Comment by pp-mo** [comment]
Pheee - ew 
Thanks @trexfeathers 😅

**Pull request merged**